component/azstorage/block_blob.go

```diff
@@ -918,6 +921,56 @@ func (bb *BlockBlob) ReadInBuffer(name string, offset int64, len int64, data []b
 	return nil
 }
 
+// ReadInBufferWithETag : Download specific range from a file to a user provided buffer, return back the current etag along with data
+func (bb *BlockBlob) ReadInBufferWithETag(name string, offset int64, len int64, data []byte) (string, error) {
+	// log.Trace("BlockBlob::ReadInBuffer : name %s", name)
+	blobClient := bb.Container.NewBlobClient(filepath.Join(bb.Config.prefixPath, name))
+	ctx, cancel := context.WithTimeout(context.Background(), max_context_timeout*time.Minute)
+	defer cancel()
+
+	opt := &blob.DownloadStreamOptions{
+		Range: blob.HTTPRange{
+			Offset: offset,
+			Count:  len,
+		},
+		CPKInfo: bb.blobCPKOpt,
+	}
+
+	dr, err := blobClient.DownloadStream(ctx, opt)
```

Comment:
Use DownloadBuffer() method here. DownloadStream() is basically a single GetBlob call which will take a lot of time if the buffer is huge.
Also, with DownloadBuffer(), you can remove the below code of copying the data to the buffer.

Reply:
If the objective to get ETag, then we can make an extra GetProperties call at the end. Since, this feature will be enabled by a flag, we should have DownloadBuffer() + GetProperties() call, which will generally be more performant than DownloadStream().

Reply:
downloadBuffer that we are using is internally using download stream only and then copies the data over to our buffer. Reason to use the stream directly is to get the etag. In downloadstream we have the responce but in download buffer we just return back data len and error so that actual responce is trimmed. If we use downloadStream directly then there is no need to make an extra getProperties call.